### PR TITLE
[feat] Adding command to import account

### PR DESCRIPTION
**Pull request by matthiashanel**
If the account is self signed, resign with the operator.

Signed-off-by: Matthias Hanel <mh@synadia.com>

manual test:
```text
> nsc delete account ACC
[ OK ] expired account "ACC"
[ OK ] deleted account
[ OK ] deleted account directory
> nsc import account --jwt import.jwt
[WARN] Validation resulted in: self-signed account JWTs shouldn't contain operator limits
[ OK ] Account ACC was successfully imported
1 job succeeded - 1 have warnings
> nsc import account --jwt import.jwt
[WARN] Validation resulted in: self-signed account JWTs shouldn't contain operator limits
[ERR ] Account already exists, overwrite with --overwrite
Error: 1 job failed - 0 job succeeded and 1 had warnings
Usage:
  nsc import account <jwt file> [flags]

Examples:
nsc import account <jwt>

Flags:
  -h, --help         help for account
  -j, --jwt string   account jwt to import
      --overwrite    overwrite existing account
      --skip         skip validation issues

Global Flags:
  -i, --interactive          ask questions for various settings
  -K, --private-key string   private key

> nsc import account --jwt import.jwt  --overwrite
[WARN] Validation resulted in: self-signed account JWTs shouldn't contain operator limits
[ OK ] Account ACC was successfully imported
1 job succeeded - 1 have warnings
> git describe account -n ACC
fatal: not a git repository (or any of the parent directories): .git
> nsc describe account -n ACC
+--------------------------------------------------------------------------------------+
|                                   Account Details                                    |
+---------------------------+----------------------------------------------------------+
| Name                      | ACC                                                      |
| Account ID                | ADRB4JJYFDLWKIMX4DH6MX2DMKA3TENJWGMNVM5ILYLZTT6BN7QIF5ZX |
| Issuer ID                 | OD5FHU4LXGDSGDHO7UNRMLW6I36QX5VPJXRQHFHMRUIKSHOPEDSHVPBB |
| Issued                    | 2020-10-28 23:07:56 UTC                                  |
| Expires                   |                                                          |
+---------------------------+----------------------------------------------------------+
| Max Connections           | Unlimited                                                |
| Max Leaf Node Connections | Unlimited                                                |
| Max Data                  | Unlimited                                                |
| Max Exports               | Unlimited                                                |
| Max Imports               | Unlimited                                                |
| Max Msg Payload           | Unlimited                                                |
| Max Subscriptions         | Unlimited                                                |
| Exports Allows Wildcards  | True                                                     |
+---------------------------+----------------------------------------------------------+
| Imports                   | None                                                     |
| Exports                   | None                                                     |
+---------------------------+----------------------------------------------------------+
> nsc list keys --all
+------------------------------------------------------------------------------------------+
|                                           Keys                                           |
+--------+----------------------------------------------------------+-------------+--------+
| Entity | Key                                                      | Signing Key | Stored |
+--------+----------------------------------------------------------+-------------+--------+
| DEMO   | OD5FHU4LXGDSGDHO7UNRMLW6I36QX5VPJXRQHFHMRUIKSHOPEDSHVPBB |             | *      |
|  ACC   | ADRB4JJYFDLWKIMX4DH6MX2DMKA3TENJWGMNVM5ILYLZTT6BN7QIF5ZX |             |        |
|  SYS   | AAYVLZJC2ULKSH5HNSKMIKFMCEHCNU5VOV5KG56IRL7ENHLBUGZ27CZT |             | *      |
|   sys  | UBVZYLLCAFMHBXBUDKKKFKH62T4AW7Q5MAAE3R3KKAIRCZNYITZPDQZ3 |             | *      |
+--------+----------------------------------------------------------+-------------+--------+

>


```

import of account and user in managed store
```
> nsc import account --jwt ../env1/nats/DEMO/accounts/SYS/SYS.jwt
[WARN] unable to push to "DEMO" - operator doesn't set an account server url or manual exchange necessary
[WARN] perform push/pull again or exchange self-signed JWT manually
[ OK ] Account SYS was successfully imported
1 job succeeded - 1 have warnings
> nsc list keys --all
+------------------------------------------------------------------------------------------+
|                                           Keys                                           |
+--------+----------------------------------------------------------+-------------+--------+
| Entity | Key                                                      | Signing Key | Stored |
+--------+----------------------------------------------------------+-------------+--------+
| DEMO   | OD5FHU4LXGDSGDHO7UNRMLW6I36QX5VPJXRQHFHMRUIKSHOPEDSHVPBB |             |        |
|  ACC   | ADRB4JJYFDLWKIMX4DH6MX2DMKA3TENJWGMNVM5ILYLZTT6BN7QIF5ZX |             | *      |
|  SYS   | AAYVLZJC2ULKSH5HNSKMIKFMCEHCNU5VOV5KG56IRL7ENHLBUGZ27CZT |             |        |
+--------+----------------------------------------------------------+-------------+--------+

> nsc import account --jwt ../env1/nats/DEMO/accounts/SYS/SYS.jwt
[ERR ] Account already exists, overwrite with --overwrite
Error: all jobs failed
Usage:
  nsc import account --jwt <account-jwt> [flags]

Examples:
nsc import account --jwt <account-jwt>

Flags:
  -h, --help         help for account
  -j, --jwt string   account jwt to import
      --overwrite    overwrite existing account
      --skip         skip validation issues, they can be edited out

Global Flags:
  -i, --interactive          ask questions for various settings
  -K, --private-key string   private key

> nsc import user --file ../env1/nats/DEMO/accounts/SYS/users/sys.jwt
[ OK ] User sys was successfully imported
> nsc list keys --all
+------------------------------------------------------------------------------------------+
|                                           Keys                                           |
+--------+----------------------------------------------------------+-------------+--------+
| Entity | Key                                                      | Signing Key | Stored |
+--------+----------------------------------------------------------+-------------+--------+
| DEMO   | OD5FHU4LXGDSGDHO7UNRMLW6I36QX5VPJXRQHFHMRUIKSHOPEDSHVPBB |             |        |
|  ACC   | ADRB4JJYFDLWKIMX4DH6MX2DMKA3TENJWGMNVM5ILYLZTT6BN7QIF5ZX |             | *      |
|  SYS   | AAYVLZJC2ULKSH5HNSKMIKFMCEHCNU5VOV5KG56IRL7ENHLBUGZ27CZT |             |        |
|   sys  | UBVZYLLCAFMHBXBUDKKKFKH62T4AW7Q5MAAE3R3KKAIRCZNYITZPDQZ3 |             |        |
+--------+----------------------------------------------------------+-------------+--------+

> nsc import user --file ../env1/keys/creds/DEMO/SYS/sys.creds
[ERR ] User already exists, overwrite with --overwrite
Error: all jobs failed
Usage:
  nsc import user --file <user-jwt/user-creds> [flags]

Examples:
nsc import account --jwt <account-jwt>

Flags:
  -f, --file string   user jwt or creds to import
  -h, --help          help for user
      --overwrite     overwrite existing user
      --skip          skip validation issues

Global Flags:
  -i, --interactive          ask questions for various settings
  -K, --private-key string   private key

> nsc import user --file ../env1/keys/creds/DEMO/SYS/sys.creds --overwrite
[ OK ] Key stored /Users/matthiashanel/test/demo/env2/keys/keys/U/BV/UBVZYLLCAFMHBXBUDKKKFKH62T4AW7Q5MAAE3R3KKAIRCZNYITZPDQZ3.nk
[ OK ] User sys was successfully imported
all jobs succeeded
> nsc list keys --all
+------------------------------------------------------------------------------------------+
|                                           Keys                                           |
+--------+----------------------------------------------------------+-------------+--------+
| Entity | Key                                                      | Signing Key | Stored |
+--------+----------------------------------------------------------+-------------+--------+
| DEMO   | OD5FHU4LXGDSGDHO7UNRMLW6I36QX5VPJXRQHFHMRUIKSHOPEDSHVPBB |             |        |
|  ACC   | ADRB4JJYFDLWKIMX4DH6MX2DMKA3TENJWGMNVM5ILYLZTT6BN7QIF5ZX |             | *      |
|  SYS   | AAYVLZJC2ULKSH5HNSKMIKFMCEHCNU5VOV5KG56IRL7ENHLBUGZ27CZT |             |        |
|   sys  | UBVZYLLCAFMHBXBUDKKKFKH62T4AW7Q5MAAE3R3KKAIRCZNYITZPDQZ3 |             | *      |
+--------+----------------------------------------------------------+-------------+--------+

 >
```

### CODE REVIEW

File: cmd/importaccount.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+func (p *ImportAccount) PreInteractive(ctx ActionCtx) error {
+	return nil
+}
+
+func (p *ImportAccount) Load(ctx ActionCtx) error {
+	return nil
+}
+
+func (p *ImportAccount) PostInteractive(ctx ActionCtx) error {
```
Comment (aricart) [inline cmd/importaccount.go:80]: Why any interactive code, if the selection of the JWT cannot happen interactively?

File: cmd/importaccount.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	SignerParams
+	fileImport
+}
+
+func (p *ImportAccount) SetDefaults(ctx ActionCtx) error {
+	p.SignerParams.SetDefaults(nkeys.PrefixByteOperator, true, ctx)
+	return nil
+}
+
+func (p *ImportAccount) PreInteractive(ctx ActionCtx) error {
```
Comment (aricart) [inline cmd/importaccount.go:72]: Add field to ask for file by prompt here (that means that file is no longer required in the flag, so need to verify in Validate)

File: cmd/importaccount.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		r.AddError("failed to decode %#q: %v", p.file, err)
+		return r, err
+	}
+	if validateAndReport(claim, p.skip, r) {
+		return r, nil
+	}
+	if ctx.StoreCtx().Store.HasAccount(claim.Name) {
+		if !p.overwrite {
+			r.AddError("Account already exists, overwrite with --overwrite")
```
Comment (aricart) [inline cmd/importaccount.go:134]: all status messages should start with a lower case to match other commands unless it is the name of some entity.

File: cmd/importaccount.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+			return r, nil
+		}
+		if old, err := ctx.StoreCtx().Store.ReadAccountClaim(claim.Name); err != nil {
+			r.AddError("Existing Account not found: %v", err)
```
Comment (aricart) [inline cmd/importaccount.go:138]: same as above

File: cmd/importaccount.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+			return r, nil
+		} else if old.Subject != claim.Subject {
+			r.AddError("Existing Account has a name collision only and does not reference the same entity "+
```
Comment (aricart) [inline cmd/importaccount.go:141]: and here

File: cmd/importaccount.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	keys, _ := ctx.StoreCtx().GetOperatorKeys()
+	for _, key := range keys {
+		if key == claim.Issuer {
+			sameOperator = true
+			break
+		}
+	}
+	if ctx.StoreCtx().Store.IsManaged() {
+		if claim.IsSelfSigned() {
+			r.AddError("Only a non managed store can import a self signed account")
```
Comment (aricart) [inline cmd/importaccount.go:157]: and here

File: cmd/importaccount.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			return r, nil
+		}
+	} else if claim.IsSelfSigned() {
+		if jwt, err := claim.Encode(p.signerKP); err != nil {
+			r.AddError("Error during encoding of self signed account jwt: %v", err)
```
Comment (aricart) [inline cmd/importaccount.go:162]: same

File: cmd/importaccount.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+			sameOperator = true
+		}
+	}
+	if !sameOperator {
+		r.AddError("Can only import account signed by same operator. Possibly update your operator")
+		return r, nil
+	}
+	sub, err := ctx.StoreCtx().Store.StoreClaim(theJWT)
+	if err != nil {
+		r.AddError("Error when storing account: %v", err)
```
Comment (aricart) [inline cmd/importaccount.go:175]: here

File: cmd/importaccount_test.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/nats-io/jwt"
+	"github.com/nats-io/nkeys"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_ImportAccountSelfSigned(t *testing.T) {
```
Comment (aricart) [inline cmd/importaccount_test.go:28]: If implementing interactive, will need one of those here as well.

File: cmd/importuser.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+	return nil
+}
+
+// returns true when blocking and not skipped
+func validateAndReport(claim jwt.Claims, force bool, r *store.Report) bool {
+	vr := jwt.ValidationResults{}
+	claim.Validate(&vr)
+	for _, i := range vr.Issues {
+		if i.Blocking {
+			r.AddError("Validation resulted in: %s", i.Description)
```
Comment (aricart) [inline cmd/importuser.go:94]: lower case `validation ...`

Comment (aricart) [reply cmd/importuser.go:94]: same for all the other messages

File: cmd/importuser.go [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+	if vr.IsBlocking(true) && !force {
+		r.AddError("Validation is blocking, skip with --skip")
+		return true
+	}
+	return false
+}
+
+func (p *ImportUser) Run(ctx ActionCtx) (store.Status, error) {
+	r := store.NewDetailedReport(true)
+	content, err := ioutil.ReadFile(p.file)
```
Comment (aricart) [inline cmd/importuser.go:108]: load should have already loaded the file...

File: cmd/importuser_test.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018-2020 The NATS Authors
```
Comment (aricart) [inline cmd/importuser_test.go:2]: new file 2020 only

File: cmd/importuser.go [hunk 13]
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2018-2020 The NATS Authors
```
Comment (aricart) [inline cmd/importuser.go:2]: 2020 only - new file

File: cmd/importaccount_test.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2018-2020 The NATS Authors
```
Comment (aricart) [inline cmd/importaccount_test.go:2]: 2020 only - new file

File: cmd/importaccount.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2018-2020 The NATS Authors
```
Comment (aricart) [inline cmd/importaccount.go:2]: 2020 only - new file

**Review by aricart — CHANGES_REQUESTED** [review]

File: cmd/importaccount.go (hunk 5 above)
Comment (matthiashanel) [reply cmd/importaccount.go:141]: fixed

File: cmd/importaccount.go (hunk 6 above)
Comment (matthiashanel) [reply cmd/importaccount.go:157]: fixed

File: cmd/importaccount.go (hunk 7 above)
Comment (matthiashanel) [reply cmd/importaccount.go:162]: fixed

File: cmd/importaccount_test.go (hunk 9 above)
Comment (matthiashanel) [reply cmd/importaccount_test.go:28]: fixed

File: cmd/importuser.go (hunk 10 above)
Comment (matthiashanel) [reply cmd/importuser.go:94]: fixed

File: cmd/importuser_test.go (hunk 12 above)
Comment (matthiashanel) [reply cmd/importuser_test.go:2]: fixed

File: cmd/importaccount_test.go (hunk 14 above)
Comment (matthiashanel) [reply cmd/importaccount_test.go:2]: fixed

File: cmd/importuser.go (hunk 13 above)
Comment (matthiashanel) [reply cmd/importuser.go:2]: fixed

File: cmd/importaccount.go (hunk 15 above)
Comment (matthiashanel) [reply cmd/importaccount.go:2]: fixed

File: cmd/importaccount.go (hunk 8 above)
Comment (matthiashanel) [reply cmd/importaccount.go:175]: fixed

File: cmd/importuser.go (hunk 11 above)
Comment (matthiashanel) [reply cmd/importuser.go:108]: moved that to Load

File: cmd/importaccount.go (hunk 4 above)
Comment (matthiashanel) [reply cmd/importaccount.go:138]: fixed

File: cmd/importaccount.go (hunk 1 above)
Comment (matthiashanel) [reply cmd/importaccount.go]: will not implement and removed

File: cmd/importaccount.go (hunk 2 above)
Comment (matthiashanel) [reply cmd/importaccount.go]: will not implement

**Review by aricart — APPROVED** [review]
LGTM

**Pull request merged**